### PR TITLE
executor: recover panic in executor.Close()

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/49223

Problem Summary:
Prevent potential goroutine leaks.

In UnionExec.Close(), if there is a panic during closing the first child executor,  the second executor will not closed.
And the same is true in BaseExecutor.Close().

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/cte_test.go [hunk 1]
```diff
@@ -53,9 +53,7 @@ func TestCTEIssue49096(t *testing.T) {
 	sql := "insert into t2 with cte1 as ( " +
 		"select c1 from t1) " +
 		"select c1 from cte1 natural join (select * from cte1 where c1 > 0) cte2 order by c1;"
-	err := tk.ExecToErr(sql)
-	require.NotNil(t, err)
-	require.Equal(t, "[executor:8175]Your query has been cancelled due to exceeding the allowed memory limit for a single SQL query. Please try narrowing your query scope or increase the tidb_mem_quota_query limit and try again.[conn=%d]", err.Error())
+	tk.MustExec(sql) // No deadlock
```
Comment (XuHuaiyu) [inline pkg/executor/cte_test.go:56]: Why change this?

Comment (wshwsh12) [reply pkg/executor/cte_test.go:56]: The error occurs in the Close() function, we should only log it.

**Review by XuHuaiyu — APPROVED** [review]

**Review by AilinKid — COMMENTED** [review]
/LGTM

**Pull request merged**